File: examples/java/Point.java
package test;

public class PointJava {

    private PointJava() {
    }

    interface Point<Self extends Point<Self>> {
        int x();

        int y();

        Self init(int x, int y);

        default Self move(int dx, int dy) {
            return init(x() + dx, y() + dy);
        }

        default boolean equal(Self s) {
            return s.x() == x() && s.y() == y();
        }
    }

    interface PointColore<Self extends PointColore<Self>> extends Point<Self> {
        int c();

        default boolean equal(Self s) {
            return s.c() == c() && Point.super.equal(s);
        }
    }

    // Recursive definition

    static class YPoint implements Point<YPoint> {
        final int x, y;

        public YPoint(int x, int y) {
            this.x = x;
            this.y = y;
        }

        @Override
        public int x() {
            return x;
        }

        @Override
        public int y() {
            return y;
        }

        @Override
        public YPoint init(int x, int y) {
            return new YPoint(x, y);
        }
    }

    static class YPointColore implements PointColore<YPointColore> {
        final int x, y, c;

        public YPointColore(int x, int y, int c) {
            this.x = x;
            this.y = y;
            this.c = c;
        }

        public int x() {
            return x;
        }

        public int y() {
            return y;
        }

        public int c() {
            return c;
        }

        @Override
        public YPointColore init(int x, int y) {
            return new YPointColore(x, y, c());
        }
    }

    static <S extends Point<S>> boolean equal(S p1, S p2) {
        return p1.equal(p2);
    }

    public static void main(String[] args) {
        final Point p1 = new YPoint(0, 0);
        final PointColore p2 = new YPointColore(0, 0, 0);

        equal(p1, p2);
    }
}

File: examples/kotlin/Point.kt
package test

interface Point<Self : Point<Self>> {
    val x: Int;
    val y: Int

    fun new(x: Int, y: Int): Self

    fun move(dx: Int, dy: Int) = new(x + dx, y + dy)
    fun equal(s: Point<Self>) = s.x == x && s.y == y
}

interface PointColore<Self : PointColore<Self>> : Point<Self> {
    val c: Int

    fun equal(s: PointColore<Self>) = s.c == c && super.equal(s)
}

// Recursive definition

class YPoint(override val x: Int, override val y: Int) : Point<YPoint> {
    override fun new(x: Int, y: Int): YPoint = YPoint(x, y)
}

class YPointColore(override val x: Int, override val y: Int, override val c: Int) :
        PointColore<YPointColore> {
    override fun new(x: Int, y: Int): YPointColore = YPointColore(x, y, c)
}

fun <P : Point<P>> List<P>.move(dx: Int, dy: Int): List<P> =
        this.map { it.move(dx, dy) }

fun <P : Point<P>> equal(p1: P, p2: P): Boolean =
        p1.equal(p2)

fun List<Point<*>>.move2(dx: Int, dy: Int): List<Point<*>> =
        this.map { it.move(dx, dy) }

fun main() {
    val p = YPoint(0, 0)
    val cp = YPointColore(0, 0, 0)
    val l = listOf(p, cp).move2(1, 2)

    cp.equal(cp)
}
